### PR TITLE
👨‍💻 DelegatedFlatStateVariant

### PR DESCRIPTION
**Pull request by sesh92**
# Purpose
The DelegatedFlatStateVariant was created to make it easier to handle certain tagged states types that were designed to be flat.

You can see the usage in tests

### CODE REVIEW

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable jest/expect-expect */
+import { AssertEqual } from "@tagged-state/testutils";
+import { DelegatedFlatStateVariant } from "./delegatedTaggedState";
+import { StateProps, StateVariant } from "./taggedState";
+
+describe("delegated tagged state", () => {
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:7]:
```suggestion
describe("DelegatedFlatStateVariant", () => {
```

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      | DelegatedFlatStateVariant<"ready", FlatVariant>
+      | DelegatedFlatStateVariant<"error", FlatVariant>;
+
+    const assert1: State = { tag: "uninit", data: {} };
+    const assert2: State = { tag: "loading", data: {} };
+    const assert3: State = { tag: "ready", data: { someVal: "value" } };
+    const assert4: State = { tag: "error", data: { error: new Error() } };
+  });
+
+  it("disallows invalid flat types", () => {
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:25]:
```suggestion
  it("disallows invalid values", () => {
```

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable jest/expect-expect */
+import { AssertEqual } from "@tagged-state/testutils";
+import { DelegatedFlatStateVariant } from "./delegatedTaggedState";
+import { StateProps, StateVariant } from "./taggedState";
+
+describe("delegated tagged state", () => {
+  it("allows valid flat types", () => {
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:8]:
```suggestion
  it("allows valid values", () => {
```

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    const assert1: State = { tag: "error", error: new Error() };
+    // @ts-expect-error: invalid data
+    const assert2: State = { tag: "error", someVal: true };
+    // @ts-expect-error: unmatched data subtype
+    const assert3: State = { tag: "error", data: { someVal: true } };
+    // @ts-expect-error: unknown tag
+    const assert4: State = { tag: "unknown", data: {} };
+  });
+
+  it("correctly derives StateProps", () => {
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:46]: This test is redundant

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable jest/expect-expect */
+import { AssertEqual } from "@tagged-state/testutils";
+import { DelegatedFlatStateVariant } from "./delegatedTaggedState";
+import { StateProps, StateVariant } from "./taggedState";
+
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:6]: We should have a test that ensures type equality between the flat state and the corresponding tagged state.

Similar to `correctly derives StateProps`, but without involving `StateProps`.

File: packages/core/src/delegatedTaggedState.spec.ts (hunk 4 above)
Comment (sesh92) [reply packages/core/src/delegatedTaggedState.spec.ts]: https://github.com/MOZGIII/tagged-state/pull/1/commits/08413d21f052fd7ed54d75a25d075a6ec1e46878

File: packages/core/src/delegatedTaggedState.spec.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    type State =
+      | StateVariant<"uninit">
+      | StateVariant<"loading">
+      | DelegatedFlatStateVariant<"ready", FlatVariant>
+      | DelegatedFlatStateVariant<"error", FlatVariant>;
+
+    // @ts-expect-error: invalid data
+    const assert1: State = { tag: "error", error: new Error() };
+    // @ts-expect-error: invalid data
+    const assert2: State = { tag: "error", someVal: true };
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.spec.ts:39]:
```suggestion
    // @ts-expect-error: missing data
    const assert1: State = { tag: "error", error: new Error() };
    // @ts-expect-error: missing data
    const assert2: State = { tag: "error", someVal: true };
```

File: packages/core/src/delegatedTaggedState.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { StateVariant } from "./taggedState";
+
+export type DelegatedFlatStateVariant<
```
Comment (MOZGIII) [inline packages/core/src/delegatedTaggedState.ts:3]: Let's move the implementation into the `taggedState.ts`

Comment (sesh92) [reply packages/core/src/delegatedTaggedState.ts:3]: I guess the `delegatedTaggedState.spec.ts` I can move into the `taggedState.spec.ts` too

File: packages/core/src/delegatedTaggedState.spec.ts (hunk 5 above)
Comment (sesh92) [reply packages/core/src/delegatedTaggedState.spec.ts:6]: https://github.com/MOZGIII/tagged-state/pull/1/commits/d8154e1df95ee0cee6d3be9dc4d32727a4c870e5

File: packages/core/src/delegatedTaggedState.ts (hunk 7 above)
Comment (sesh92) [reply packages/core/src/delegatedTaggedState.ts]: https://github.com/MOZGIII/tagged-state/pull/1/commits/d8154e1df95ee0cee6d3be9dc4d32727a4c870e5

**Pull request merged**